stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,269 @@ public func gradient<T, U, V, R>(
   return { x, y, z in gradient(at: x, y, z, in: f) }
 }
 
+//===----------------------------------------------------------------------===//
+// Type-erased `AnyDerivative`
+//===----------------------------------------------------------------------===//
+
+internal protocol _AnyDerivativeBox {
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool
+  func _isNotEqual(to other: _AnyDerivativeBox) -> Bool
+
+  // `AdditiveArithmetic` requirements.
+  static var _zero: _AnyDerivativeBox { get }
+  static var _cotangentZero: _AnyDerivativeBox { get }
+  func _adding(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _subtracting(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  // `Differentiable` requirements.
+  var _allDifferentiableVariables: _AnyDerivativeBox { get }
+  func _moved(along direction: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _tangentVector(from cotangent: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  /// The underlying base value.
+  var _base: Any { get }
+
+  /// Returns the underlying value unboxed to the given type, if possible.
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+
+  /// Returns true if the underlying value has type `AnyDerivative.OpaqueZero`.
+  func _isOpaqueZero() -> Bool
+}
+
+internal struct _ConcreteDerivativeBox<T> : _AnyDerivativeBox
+  where T : Differentiable, T.TangentVector == T,
+        T.AllDifferentiableVariables == T,
+        // NOTE: The requirement below should be defined on `Differentiable`.
+        // But it causes a crash due to generic signature minimization bug.
+        T.CotangentVector == T.CotangentVector.AllDifferentiableVariables
+{
+  /// The underlying base value.
+  var _baseDerivative: T
+
+  public init(_ base: T) {
+    self._baseDerivative = base
+  }
+
+  /// The underlying base value, type-erased to `Any`.
+  var _base: Any {
+    return _baseDerivative
+  }
+
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+  {
+    return (self as _AnyDerivativeBox as? _ConcreteDerivativeBox<U>)?
+      ._baseDerivative
+  }
+
+  func _isOpaqueZero() -> Bool {
+    return T.self == AnyDerivative._OpaqueZero.self
```

Comment:
Can this be implemented as `_unboxed(to: AnyDerivative.OpaqueZero.self) != nil`? If so, remove this requirement from the protocol and define this method as a protocol extension method. This will help reduce the size of the protocol (fewer customization points).